### ISSUE TITLE
Recognition with KNN is not really good 

### ISSUE THREAD

**Issue by khanhnt** [deepinsight/insightface#775]
Hi Insightface team, thank you for sharing the code. I have tried to develop a recoginition demo using KNN. But I found the results is not good as I have done with face_recoginition of Dlib. The distance between faces and those in the model trained by archface is alway higher than face_recognition lib (mean that less confidence). Could you pls help me to review if I did smt wrong ? 
Here is the code I use to train : 

    `
    X = []
    y = []
    parser = argparse.ArgumentParser(description='face model test')
    parser.add_argument('--image-size', default='112,112', help='')
    parser.add_argument('--model', default='./deploy/model-r100-ii/model,0', help='path to load model.')
    parser.add_argument('--ga-model', default='./deploy/gamodel-r50/model,0', help='path to load model.')
    parser.add_argument('--gpu', default=0, type=int, help='gpu id')
    parser.add_argument('--det', default=0, type=int,
                        help='mtcnn option, 1 means using R+O, 0 means detect from begining')
    parser.add_argument('--flip', default=0, type=int, help='whether do lr flip aug')
    parser.add_argument('--threshold', default=1.24, type=float, help='ver dist threshold')
    args = parser.parse_args()
    model = face_model.FaceModel(args)
    today = date.today().strftime("%d%m%Y")

    # Loop through each person in the training set
    i = 0;
    for class_dir in os.listdir(train_dir):
        i += 1
        print(str(i) + '------------------' + class_dir)
        if not os.path.isdir(os.path.join(train_dir, class_dir)):
            continue

        # Loop through each training image for the current person
        for img_path in image_files_in_folder(os.path.join(train_dir, class_dir)):
            img = cv2.imread(img_path)
            img = model.get_input(img)
            print('Running inference... %s', img_path)
            if  img is None or len(img) == 0 :
                 print('Cannot calculate ', img_path)
                 continue
             start = time.time()
             _face_description = model.get_feature(img)
             print("Face extract took {} seconds.".format(time.time() - start))
             X.append(_face_description)
             y.append(class_dir)
    clf = neighbors.KNeighborsClassifier(n_neighbors=n_neighbors, algorithm=knn_algo, weights='distance')
    clf.fit(X, y)
    return clf
`
and the code to recognize is following: 

      ```

       video_capture = cv2.VideoCapture(video_path + filename)
       ret, frame = video_capture.read()
       img = self.face_describer.get_input(frame)
       _face_description = self.face_describer.get_feature(img)
       _face_description = np.expand_dims(_face_description, axis=0)
       # Use the KNN model to find the best matches for the test face
       closest_distances = clf.kneighbors(_face_description, n_neighbors=5)
       are_matches = [closest_distances[0][i][0] <= distance_threshold for i in range(len(closest_distances[0]))]
       # Predict classes and remove classifications that aren't within the threshold
       return [(closest_distances[0][0][0], pred, loc) if rec else (closest_distances[0][0][0], "unknown", loc) for pred, loc, rec in zip(clf.predict(_face_description), X_face_locations, are_matches)]
```

Thanks.

**Comment by syedmustafa54** [comment]
I have tried it but the same issue not that accurate so I have played with `distance_threshold` its 0.9 so detecting little bit accurate 
@khanhnt  share you full code if someone wants to try and suggest you I have Implemented this same code and tweak some var getting a decent accuracy not as fac_rec so if you find anything which is giving you more accuracy please update me I am also trying the same thing with CLF file and apache  solr

**Comment by khanhnt** [comment]
I think 0.9 for distance _threshold is a little bit high. Even with the face_recogition of dlib, they set the defaul thresold to 0.6 (the value can be changed depending on the dataset). As I know the arcface loss will well separate different features, that mean the distance will be smaller then.

**Comment by syedmustafa54** [comment]
Yes if I am giving 0.6 it's saying unknown when I am increasing to 0.9 it's showing properly so if you find how to improve accuracy please let me know

**Comment by tranvanhoa533** [comment]
which type of distance did you use ?

**Comment by khanhnt** [comment]
> which type of distance did you use ?

I used the defaul distance of sklearn's KNN, that mean euclidean distance

**Comment by tranvanhoa533** [comment]
If you use l2, you should normalize embedding first and threshold for
distance is usually high ( about 1.0). If you use cosine, the threshold is
smaller (about 0.5)

Trần Văn Hòa
Khoa KH&KT máy tính - ĐH bách khoa TPHCM
Phone: 01662357741


Vào Th 6, 19 thg 7, 2019 vào lúc 08:24 khanhnt <notifications@github.com>
đã viết:

**Comment by syedmustafa54** [comment]
@khanhnt  did you solve this using cosine as @tranvanhoa533  said if yes can you share your feedback on it.
@tranvanhoa533 are you saying about cosine distance?
